docs(website): rewrite docs for proxing to external API (#1461)
commercetools/merchant-center-application-kit: website/src/content/main-concepts/proxy-to-external-api.mdx

Since your backend server is hosted on a different domain, you can't send authenticated requests to your server from your Custom Application. Well, you can but you don't know if the request comes from an authenticated user or not.

# Proxying requests to the external API

To be able to validate that the request comes from an authenticated user, the Merchant Center API Gateway provides an endpoint that must be used to connect to your external API in a secured manner.

```
/proxy/forward-to
```

> Requests to that endpoint should additionally pass the following HTTP headers:

Reviewer: Should as that it's optional and not required?

Author: At the end it does not matter much. If you use `v2`, the `X-Project-Key` header is validated. If you use the complementary tools that we built (which are documented here), the headers are automatically set.

- `Accept-version: v2`: See [Versioning](#versioning).
- `X-Forward-To: <url>`: The URL to your external API.
- `X-Project-Key: <project-key>`: The key of the project currently being used by the Custom Application. The Merchant Center API Gateway will perform a validation check to ensure that the user has access to the project, then forward the request to your server only if the check was successful.

> To facilitate the usage of the [built-in Apollo client and the SDK actions](/main-concepts/data-fetching), we provide some utility helpers to abstract away the request configuration for the HTTP headers listed above.

Reviewer 2: utility helpers feels a bit double :)

```suggestion
To facilitate the usage of the [built-in Apollo client and the SDK actions](/main-concepts/data-fetching), we provide some helpers to easily integrate with the request configuration for the HTTP headers listed above.
```

## Usage for Apollo

> We can leverage the `context` option for Apollo queries to change a bit how the request is configured. The `@commercetools-frontend/application-shell` package now exposes an utility function to configure the Apollo context for the `/proxy/forward-to` usage.

Reviewer 2: ```suggestion
We can leverage the `context` option for Apollo queries to adjust how request is configured and executed. The `@commercetools-frontend/application-shell` package now exposes an utility function to configure the Apollo context for the `/proxy/forward-to` usage.
```

```js
import { useQuery } from 'react-apollo';
import { createApolloContextForProxyForwardTo } from '@commercetools-frontend/application-shell';
const Fetcher = () => {
  const { loading, data, error } = useQuery(MyQuery, {

>     context: createApolloContextForProxyForwardTo({

Reviewer 2: This is technically correct but I would suggest to set it up like this.

I would suggest that people use a `forward-to-apollo-contexts.js`-module

```js
const customServiceApolloContext = createApolloContextForProxyForwardTo({
  uri: 'https://my-custom-app.com/graphql',
})

export { customServiceApolloContext } 
```

to then more convienently do:

```js
import { customServiceApolloContext } from './forward-to-apollo-contexts'

const { loading, data, error } = useQuery(MyQuery, {
   context: customServiceApolloContext
})
```

cause I would assume they have many queries and mutations.

Author: It's not that simple. Most likely the URL is defined in the `env.json`, so it's passed through the app context.

I could maybe update the example with that, if that makes it clearer.

Defining the config statically only works if you hardcode the URL.

Reviewer 2: Yeah makes sense. I wonder though how many have variable endpoints. I'd make the docs work for the majority center best practices around that and no have minior edge cases influence the majority.

Reviewer 2: If we plan support variable endpoints then we should maybe have a `useProxyForwardToContext`-hook which people can use and which encapsulates a respective or configurable env.json value

```js
const forwardToCustomServiceContext = useProxyForwardToContext({ envKey: 'foo' })
const { loading, data, error } = useQuery(MyQuery, {
   context: forwardToCustomServiceContext
})
```

Author: In general you would always want to use variables. I also saw a couple of cases where `env.json` has been used. Which is actually how it was intended to be used.

I understand your point of encapsulating things but it just gets hard to find the balance to accommodate different use cases.

I think I would just update the doc example to read the value from the app context. People can then decide on their own how to access the URL.

Reviewer 2: Makes sense. Thanks for taking things into consideration!

actions.forwardTo.del({ uri: 'https://my-custom-app.com/graphql' });
actions.forwardTo.head({ uri: 'https://my-custom-app.com/graphql' });
actions.forwardTo.post({
  uri: 'https://my-custom-app.com/graphql',
  payload: { say: 'Hello' },
});
```

<Info>

As a reminder, action creators can be be dispatched using the `useAsyncDispatch` React hook. You can [check out an example](https://github.com/commercetools/merchant-center-application-kit/blob/e9548422ca8580a5846f80a250ccac043806ee3c/playground/src/components/echo-server/echo-server.js#L28-L52) in the Playground application.

</Info>

# Authenticating requests from the external API

When a valid request is sent via the `/proxy/forward-to` endpoint, the Merchant Center API Gateway forwards the request to the external API with an `Authorization: Bearer <token>` HTTP header.

The bearer token is a short-living JSON Web Token (JWT) that is used exclusively for the authorization exchange between the Merchant Center API Gateway and the external API. The token is valid for **60 seconds**.

The external API must verify that the token is valid by [using the JSON Web Key Set endpoint](#using-json-web-key-set-endpoint).

<Warning>

> It's imperative that the external API is securely protected and that only users with access to the project are able to connect to the API.

Reviewer 2: ```suggestion
It's imperative that the external API is securely protected and that only users with access to the project should able to connect to the API.
```

</Warning>

## Validating the JSON Web Token

> To facilitate the implementation of validating the JWT from the `Authorization: Bearer <token>` HTTP header, we provide a package with built-in functions and helpers to perform the heavy work.

Reviewer 2: ```suggestion
To facilitate the implementation of validating the JWT from the `Authorization: Bearer <token>` HTTP header, we provide a package with built-in functions and helpers to facilitate the heavy lifting.
```

Author: We're repeating `facilitate`. We should use another word

Reviewer 2: True, happy to leave it as is then.

```console noPromptLines="2"
npm install --save @commercetools-backend/express

yarn add @commercetools-backend/express
```

### Usage for Express.js

The package includes an async Express.js middleware that performs the token validation and assigns a `session` object to the `request` object.

```ts
type TSession = {
  userId: string;
  projectKey: string;
};
```

The middleware requires some options:

- `audience` (_string_): The public-facing URL of your API server. The value should only contain the origin URL (protocol, hostname, port), the request path is inferred from the incoming request.

  For example, given the external API is hosted at `https://my-api-server.com`, the `audience` value must be set to `https://my-api-server.com`.

- `issuer` (_string_): Either a cloud identifier or a valid URL to the Merchant Center API Gateway. The cloud identifier maps to the Merchant Center API URL of the related [cloud region](/main-concepts/api-gateway#cloud-regions).

>   | Cloud Identifier | Merchant Center API Gateway URL |

Reviewer 2: Nit: format the table to match rows and content maybe.

Author: I would prefer it like this, otherwise it gets messy when looking at diffs.

  | --- | --- |
  | `gcp-au` | `https://mc-api.australia-southeast1.gcp.commercetools.com` |
  | `gcp-eu` | `https://mc-api.europe-west1.gcp.commercetools.com` |
  | `gcp-us` | `https://mc-api.us-central1.gcp.commercetools.com` |
  | `aws-fra` | `https://mc-api.eu-central-1.aws.commercetools.com` |
  | `aws-ohio` | `https://mc-api.us-east-2.aws.commercetools.com` |

> - `inferIssuer` (_boolean_): Determines whether the issuer should be inferred from the custom request HTTP header `x-mc-api-cloud-identifier` which is sent by the Merchant Center API Gateway when forwarding the request. This might be useful in case the server is used in multiple regions. **Default: `false`**

Reviewer 2: The golden option.

- `jwks` (_object_): See [options of `jwks-rsa`](https://github.com/auth0/node-jwks-rsa/).

> Finally, you can use the middleware as following:

Reviewer 2: Finally is a bit odd cause we just explain above :)

```suggestion
You can use the middleware as following:
```

    audience: 'https://my-api-server.com',
    issuer: CLOUD_IDENTIFIERS.GCP_EU,
  })
);
app.use((request, response, next) => {
  // `request.session` contains the useful information
});
```

### Usage for Serverless Functions

The package also exposes an async `createSessionAuthVerifier` factory function that can be used directly instead of the middleware. In fact, the middleware is just a thin wrapper around this function.

The options to configure the function are the same as the Express.js middleware.

Below is an example of validating the JWT for Google Cloud Functions:

```js highlightLines="2,6-9,12-17"
const {
  createSessionAuthVerifier,
  CLOUD_IDENTIFIERS,
} = require('@commercetools-backend/express');

const sessionAuthVerifier = createSessionAuthVerifier({
  audience: 'https://my-api-server.com',
  issuer: CLOUD_IDENTIFIERS.GCP_EU,
});

exports.handler = async function (request, response) {
  try {
    await sessionAuthVerifier(request, response);
  } catch (error) {
    response.status(401).send(JSON.stringify({ message: 'Unauthorized' }));
    return;
  }
  // `request.session` contains the useful information
};
```

## Using JSON Web Key Set endpoint

> The Merchant Center API Gateway exposes a `/.well-known/jwks.json` endpoint that can be used to validate JWTs, as per Open ID Connect (OIDC) spec. The endpoint provides a rotating public key used to verify the JWT signature.

Reviewer 2: ```suggestion
The Merchant Center API Gateway exposes a `/.well-known/jwks.json` endpoint that can be used to validate JWTs, as defined Open ID Connect (OIDC) specification. The endpoint provides a rotating public key used to verify the JWT signature.
```

Additionally, there is also a `/.well-known/openid-configuration` discovery endpoint.

You can read more about JSON Web Key Set [here](https://auth0.com/docs/jwks) and [here](https://tools.ietf.org/html/rfc7517).

<Info>

> If you use the `@commercetools-backend/express` package, you don't need to do anything in this regard.

Reviewer 2: ```suggestion
The setup described above is not needed when using the `@commercetools-backend/express` package which contains all needed setup already.
```

</Info>

# Versioning

The `/proxy/forward-to` endpoint is versioned using the `Accept-version` HTTP header.

Versions follow an incremental number in the format `v1`, `v2`, etc.

<Info>

**Important**: The default version of the API may change in the future. If you're building a Custom Application and care about the stability of the API, be sure to request a specific version in the `Accept-version` HTTP header.

</Info>

## Current and deprecated versions

The current version refers to the **default** version used by the API, if no `Accept-version` header is provided.

When we introduce a new version, the default version enters a **deprecation period**, after which the version is marked as deprecated and the new version becomes the **default** version.

| Feature maturity     | Deprecation period |
| -------------------- | ------------------ |
| Beta                 | 4 months           |
| General Availability | 12 months          |

> The following table lists all the supported versions and their deprecation period:

Reviewer 2: ```suggestion
The following table lists all the supported versions and their potential respective deprecation start period:
```

| Version  | Is default |  Deprecated from  |

> | -------- | ---------- | ---------- |

Reviewer 2: Nit: format.
